.circleci/config.yml

```diff
@@ -46,6 +49,11 @@ jobs:
             name: Build Docker images
             working_directory: /ichnaea
             command: |
+                if [ "${DOCKER_USER}" == "" ] || [ "${DOCKER_PASS}" == "" ]; then
+                  echo "Skipping Login to Dockerhub, credentials not available."
+                else
+                  docker login --username "${DOCKER_USER}" --password "${DOCKER_PASS}"
```

Comment:
Is doing this still necessary, or is it redundant now that you have the `docker: auth:` block ?

Reply:
It's a good question. I'm not sure what ``docker login`` does under the hood in CircleCI. Here's the [docker login](https://docs.docker.com/engine/reference/commandline/login/) docs, there are a few ways to store credentials.

When I use it locally with personal credentials, ``~/.docker/config.json`` says I'm using ``"credsStore": "desktop"``. This suggests to me that Docker Desktop is storing credentials in memory. When I run ``docker login`` twice, it says "Authenticating with existing credentials..."

The [CircleCI config reference](https://circleci.com/docs/2.0/configuration-reference/#docker) doesn't say how things are done. I could imagine:

1. You just login once, and the rest of the docker commands use the same authenticated session
2. You need to login at the level of fetching the docker image, and again for build steps using the docker image
3. You need to login to fetch the image, and again for _each_ step that runs a docker command.

Some experimentation may be needed to figure out which applies. I'd be quite happy if one login only was needed. I believe I can test this tomorrow, to see if I can tag and push a release with just the ``docker: auth`` block.

I tried to use ``--password`` instead of ``password-stdin`` here, but when I try it locally, I get a warning:

```
WARNING! Using --password via the CLI is insecure. Use --password-stdin.
```

So, I'll at least want to change that.

Reply:
I tried just the ``docker: auth`` block in #1372. It worked as a PR, but I wasn't authenticated for the push when merging to master. I think we need at least one ``docker login`` in the build steps as well.